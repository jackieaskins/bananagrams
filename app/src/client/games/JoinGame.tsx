import {
  Box,
  Button,
  Link as ChakraLink,
  Heading,
  Stack,
} from "@chakra-ui/react";
import { useCallback, useState } from "react";
import {
  Link as ReactRouterLink,
  useNavigate,
  useParams,
} from "react-router-dom";
import { GameLocationState } from "./types";
import { useSavedUsername } from "@/client/LocalStorageContext";
import ErrorAlert from "@/client/alerts/ErrorAlert";
import CheckboxField from "@/client/forms/CheckboxField";
import InputField from "@/client/forms/InputField";
import CenteredLayout from "@/client/layouts/CenteredLayout";
import { socket } from "@/client/socket";
import { ClientToServerEventName } from "@/types/socket";

type JoinGameParams = {
  gameId: string;
};

export default function JoinGame(): JSX.Element {
  const navigate = useNavigate();
  const { gameId } = useParams<JoinGameParams>() as JoinGameParams;

  const [savedUsername, setSavedUsername] = useSavedUsername();
  const [username, setUsername] = useState(savedUsername);
  const [isSpectator, setIsSpectator] = useState(false);
  const [error, setError] = useState("");
  const [isJoiningGame, setIsJoiningGame] = useState(false);

  const onSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement>): void => {
      event.preventDefault();

      setIsJoiningGame(true);
      setSavedUsername(username);

      socket.emit(
        ClientToServerEventName.JoinGame,
        { gameId, username, isSpectator },
        (error, gameInfo) => {
          if (gameInfo) {
            const locationState: GameLocationState = {
              isInGame: true,
              gameInfo,
            };

            navigate("..", { state: locationState, relative: "path" });
          } else {
            setError(error?.message ?? "Unable to join game");
            setIsJoiningGame(false);
          }
        },
      );
    },
    [gameId, isSpectator, navigate, setSavedUsername, username],
  );

  return (
    <CenteredLayout>
      <Box as="form" autoComplete="off" onSubmit={onSubmit}>
        <Stack spacing={6}>
          <Heading as="h1" textAlign="center">
            Join game
          </Heading>

          <Stack>
            <ErrorAlert visible={!!error}>{error}</ErrorAlert>

            <InputField
              label="Username"
              name="username"
              isRequired
              value={username}
              setValue={setUsername}
              autoComplete={username}
            />

            <CheckboxField
              name="isSpectator"
              checked={isSpectator}
              setChecked={setIsSpectator}
              helperText="Note: You will automatically be made a specator if you join while a game is in progress"
            >
              Join as spectator
            </CheckboxField>
          </Stack>

          <Button
            colorScheme="blue"
            isLoading={isJoiningGame}
            loadingText="Joining game"
            type="submit"
          >
            Join game
          </Button>

          <ChakraLink
            as={ReactRouterLink}
            to="../../.."
            textAlign="center"
            relative="path"
          >
            Go home
          </ChakraLink>
        </Stack>
      </Box>
    </CenteredLayout>
  );
}
