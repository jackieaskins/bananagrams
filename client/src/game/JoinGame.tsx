import { Card, Col, Row } from 'antd';
import { Link, useParams } from 'react-router-dom';

import JoinGameForm from './JoinGameForm';

type JoinGameParams = {
  gameId: string;
};

const JoinGame = (): JSX.Element => {
  const { gameId } = useParams<JoinGameParams>();

  return (
    <Row css={{ height: '100vh' }} align="middle" justify="center">
      <Col xs={23} sm={21} md={16} lg={12}>
        <Card bordered={false}>
          <h1 css={{ textAlign: 'center' }}>Join game</h1>

          <JoinGameForm gameId={gameId} showGameIdField={false} />

          <div css={{ width: '100%', textAlign: 'center' }}>
            <Link to="/">Go home</Link>
          </div>
        </Card>
      </Col>
    </Row>
  );
};

export default JoinGame;
