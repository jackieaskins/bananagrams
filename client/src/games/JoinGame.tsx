import CenteredLayout from "../layouts/CenteredLayout";
import JoinGameForm from "./JoinGameForm";

export default function JoinGame(): JSX.Element {
  return (
    <CenteredLayout>
      <JoinGameForm />
    </CenteredLayout>
  );
}
