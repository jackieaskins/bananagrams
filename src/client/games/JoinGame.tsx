import {
  Box,
  Button,
  Link as ChakraLink,
  Heading,
  Stack,
} from "@chakra-ui/react";
import { useCallback, useState } from "react";
import {
  Link as ReactRouterLink,
  useNavigate,
  useParams,
} from "react-router-dom";
import { useSavedUsername } from "../LocalStorageContext";
import ErrorAlert from "../alerts/ErrorAlert";
import CheckboxField from "../forms/CheckboxField";
import InputField from "../forms/InputField";
import CenteredLayout from "../layouts/CenteredLayout";
import { useSocket } from "../socket/SocketContext";
import { GameInfo, GameLocationState } from "./types";

type JoinGameParams = {
  gameId: string;
};

export default function JoinGame(): JSX.Element {
  const { socket } = useSocket();
  const navigate = useNavigate();
  const { gameId } = useParams<JoinGameParams>() as JoinGameParams;

  const [savedUsername, setSavedUsername] = useSavedUsername();
  const [username, setUsername] = useState(savedUsername);
  const [isSpectator, setIsSpectator] = useState(false);
  const [error, setError] = useState("");
  const [isJoiningGame, setIsJoiningGame] = useState(false);

  const onSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement>): void => {
      event.preventDefault();

      setIsJoiningGame(true);
      setSavedUsername(username);

      socket.emit(
        "joinGame",
        { gameId, username, isSpectator },
        (error: Error | null, gameInfo: GameInfo) => {
          if (error) {
            setError(error.message);
            setIsJoiningGame(false);
          } else {
            const locationState: GameLocationState = {
              isInGame: true,
              gameInfo,
            };

            navigate("..", { state: locationState, relative: "path" });
          }
        },
      );
    },
    [gameId, isSpectator, navigate, setSavedUsername, socket, username],
  );

  return (
    <CenteredLayout>
      <Box as="form" autoComplete="off" onSubmit={onSubmit}>
        <Stack spacing={6}>
          <Heading as="h1" textAlign="center">
            Join game
          </Heading>

          <Stack>
            <ErrorAlert visible={!!error}>{error}</ErrorAlert>

            <InputField
              label="Username"
              name="username"
              isRequired
              value={username}
              setValue={setUsername}
              autoComplete={username}
            />

            <CheckboxField
              name="isSpectator"
              checked={isSpectator}
              setChecked={setIsSpectator}
              helperText="Note: You will automatically be made a specator if you join while a game is in progress"
            >
              Join as spectator
            </CheckboxField>
          </Stack>

          <Button
            colorScheme="blue"
            isLoading={isJoiningGame}
            loadingText="Joining game"
            type="submit"
          >
            Join game
          </Button>

          <ChakraLink
            as={ReactRouterLink}
            to="../../.."
            textAlign="center"
            relative="path"
          >
            Go home
          </ChakraLink>
        </Stack>
      </Box>
    </CenteredLayout>
  );
}
