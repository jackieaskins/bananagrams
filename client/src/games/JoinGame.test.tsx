import React from 'react';
import { shallow } from 'enzyme';
import JoinGame from './JoinGame';

jest.mock('react-router-dom', () => ({
  useParams: () => ({
    gameId: 'gameId',
  }),
}));

describe('<JoinGame />', () => {
  test('renders properly', () => {
    expect(shallow(<JoinGame />)).toMatchSnapshot();
  });
});
