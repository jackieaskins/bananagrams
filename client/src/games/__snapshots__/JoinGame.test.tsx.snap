// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JoinGame /> renders properly 1`] = `
<CenteredLayout>
  <JoinGameForm />
</CenteredLayout>
`;
